src/chain.h

```diff
@@ -147,6 +148,8 @@ class CBlockIndex
 
     //! (memory only) Sequential id assigned to distinguish order in which blocks are received.
     uint32_t nSequenceId;
+    //! (memory only) The time this node first heard about blocks
+    static std::map<const CBlockIndex*, int64_t> firstSeenTime;
```

Comment:
Not a bad idea. But I think that this means that the destructor of CBlockIndex needs to be introduced that removes itself from this map to avoid a memory leak.